BUG: Fix some PeriodIndex resampling issues

winklerand opened the pull request:
closes #15944 
xref partially #12884
closes #13083
closes #13224

This PR addresses some of the issues related to PeriodIndex resampling.
As I'm new to the pandas codebase, I appreciate any advice.

winklerand committed:
CLN: move PeriodIndex binning code to TimeGrouper

Simplifies PeriodIndexResampler and makes it more similar to Resampler,
DatetimeIndexResampler and TimedeltaIndexResampler.

winklerand committed:
TST/CLN: raise error when resampling with on= or level= selection

Added tests to cover all code paths, moved the check to _convert_obj() and
removed then-redundant check from _upsample() method.

winklerand committed:
BUG: resampling PeriodIndex now returns PeriodIndex (GH 12884, 15944)

Exceptions:
- force conversion to DatetimeIndex by kind='timestamp' param
- if loffset is given, convert to timestamps in any case

winklerand committed:
BUG: OHLC-upsampling of PeriodIndex now returns DataFrame (GH 13083)

winklerand committed:
BUG: enable resampling with NaT in PeriodIndex (GH 13224)

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2252,12 +2278,17 @@ def test_selection(self):
-
-        with pytest.raises(NotImplementedError):
-            df.resample('2D', level='d')
+        for freq in ['H', '12H', '2D', 'W']:
```
these tests are more clear with parametrize 
but we still inherit from TestCase so can't change this

but you can start a new test class that does this 

again if it's more clear

winklerand replied:
not sure if I understood correctly - pushed 48e324f as an example, are you thinking along those lines of using pytest fixtures/mark.parametrize decorators?

jreback replied:
commented directly on the commit

jreback commented:
lmk when you need review.

winklerand commented:
ok, gave parametrized tests a shot in 68f02f7.
In my opinion, those are much more readable - it's clear at first sight which combinations of parameters/inputs are tested.

If you agree, I'd be happy to help more of the tests in TestPeriodIndex. I'd like to tackle some more issues with PeriodIndex in a follow-up PR (```convention``` parameter, maybe drop ```kind``` parameter altogether, ...) which would require better coverage of all parameter combinations/edge cases.

winklerand commented on pandas/tests/test_resample.py:
```diff
@@ -2799,89 +2693,182 @@ def test_evenly_divisible_with_no_extra_bins(self):
+        return DataFrame({'value': np.arange(len(index))}, index=index)
+
+    @pytest.fixture(params=[Series, DataFrame], scope='class')
+    def pandas_obj(self, request, index):
```
there is probably a more elegant way to create such "meta-fixtures"? ;-)

jreback replied:
I would do it more like this

```
@pytest.fixture(params=[PI, DTI, TDI])
def index(klass):
     # construct an index
     return klass(.....)

@pytest.fixture
def series(index):
    return Series(.....)

@pytest.fixture
def frame(index):
     return DataFrame(....)
```

it will cartesian product the indexes separately for series/frame which would now be the fixture arg to a test function.

winklerand replied:
tried another approach by keeping the existing inheritance structure - we can inherit fixtures defined in the ```Base``` class, ```index``` fixture defined in `Base` dispatches to the subclass-specific type via the existing ```_index_factory```. If you want to take a look:
https://github.com/winklerand/pandas/blob/bbf01b312a2d8be0b3c7681b9a46a039f33fdcc/pandas/tests/test_resample_pytest.py

Seems easier to me than parametrizing the index fixture, there's a lot of index type-specific test logic.

Should we move out this discussion to another issue or PR? This looks like a more general issue on how to (gradually) transition to pytest.

jreback replied:
NO

we are removing the base classes soon

jreback replied:
sorry u meant the Base class and not TestCase base class

in any event it's not more clear to me

winklerand replied:
ok, was maybe a bit confusing ;-)

pushed e497fc2, should illustrate what I meant better: basically move fixtures to a Base_pytest(object) class which can be inherited by TestPeriodIndex_pytest(Base_pytest), and DatetimeIndex_pytest(Base_pytest). This resembles the existing test organization structure, but allows for parametrization using the common fixtures.

MaximilianR commented on pandas/core/resample.py:
```diff
@@ -793,53 +793,34 @@ class PeriodIndexResampler(DatetimeIndexResampler):
+
+        if self.loffset is not None:
+            if self.kind == 'period':
+                print('Warning: loffset -> convert PeriodIndex to timestamps')
```
Use `warnings.warn` rather than `print`

jreback replied:
was this here before? we don't warn on things like this.

winklerand replied:
a similar print-style warning is currently in master: https://github.com/pandas-dev/pandas/blob/master/pandas/core/resample.py#L800:L802

MaximilianR replied:
I defer to @jreback

winklerand replied:
removed the warning in 8a4182e - any objections?

winklerand committed:
CLN: remove warning on falling back to tstamp resampling with loffset

jreback commented on pandas/core/resample.py:
```diff
@@ -857,22 +836,16 @@ def _downsample(self, how, **kwargs):
+            return self._groupby_and_aggregate(how, grouper=self.grouper)
         elif is_superperiod(ax.freq, self.freq):
+            if how == 'ohlc':
+                # upsampling to subperiods is handled as an asfreq, which works
```
can you add the issue reference here

winklerand replied:
done

jreback commented on pandas/core/resample.py:
```diff
@@ -1269,6 +1239,58 @@ def _get_time_period_bins(self, ax):
+                            'an instance of %r' % type(ax).__name__)
+
+        memb = ax.asfreq(self.freq, how=self.convention)
+        # NaT handling as in pandas._lib.lib.generate_bins_dt64()
```
blank line

winklerand replied:
done

jreback commented on pandas/core/resample.py:
```diff
@@ -1269,6 +1239,58 @@ def _get_time_period_bins(self, ax):
+        nat_count = 0
+        if memb.hasnans:
+            import warnings
+            with warnings.catch_warnings():
```
you don't need this

``memb._isnan`` already has this mask.

jreback commented on pandas/core/resample.py:
```diff
@@ -1269,6 +1239,58 @@ def _get_time_period_bins(self, ax):
+
+        i8 = memb.asi8
+        freq_mult = self.freq.n
+        # when upsampling to subperiods, we need to generate enough bins
```
blank line

winklerand replied:
done

jreback commented on pandas/core/resample.py:
```diff
@@ -1269,6 +1239,58 @@ def _get_time_period_bins(self, ax):
+        freq_mult = self.freq.n
+        # when upsampling to subperiods, we need to generate enough bins
+        expected_bins_count = len(binner) * freq_mult
+        i8_extend = expected_bins_count - (i8[-1] - i8[0])
```
this will fail if ``not len(memb)``

winklerand replied:
Some lines above (after masking out NaTs), we check for ```if not len(memb)``` and return and empty index, so this case should be caught.
Maybe put in an ```assert len(memb)``` to make it explicit?

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2777,6 +2694,211 @@ def test_evenly_divisible_with_no_extra_bins(self):
+                                 freq=self._index_fixture_freq)
+
+    @pytest.fixture(params=[Series, DataFrame], scope='class')
+    def pandas_obj(self, request, index):
```
make this 2 fixtures, series and frame, much simpler, and easier to grok.

jreback replied:
you can *then* make another fixture like ``series_and_frame`` if you want, though I find that more confusing, rather have you use params.

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2777,6 +2694,211 @@ def test_evenly_divisible_with_no_extra_bins(self):
         assert_frame_equal(result, expected)
 
 
+def assert_series_or_frame_equal(result, expected):
```
this is prob ok, but move to ``util.testing``

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2777,6 +2694,211 @@ def test_evenly_divisible_with_no_extra_bins(self):
+    pass
+
+
+class TestPeriodIndex_pytest(Base_pytest):
```
I would rather name these ``TestPeriodIndex(Base)`` and rename the original tests I think (and then we should move them in other PR's)

jreback reviewed (changes requested):
looks pretty good. I have to do another look on the actual tests.

winklerand committed:
CLN: use memb._isnan for NaT masking

winklerand committed:
DOC: added issue reference for OHLC resampling

winklerand committed:
STYLE: added blank lines

winklerand committed:
TST: convert to parametrized tests / pytest idiom

winklerand committed:
CLN/TST: call assert_almost_equal() when comparing Series/DataFrames

Removed previously defined helper method assert_series_or_frame_equal(),
which behaved identical to assert_almost_equal()

winklerand commented:
Rebased, so we can put the parametrized test in the existing test classes (which just inherit from ```object```).
Thanks for the great feedback!

jreback commented:
cc @MaximilianR can you have a look.

MaximilianR commented:
Looks great. Fixes a couple of important issues. 

Whether or not we reveal `kind` on resamples is an open question - I think there already enough tools for switching between DateTimeIndex and PeriodIndex that we don't need more in `resample`. 

But this definitely makes things better! Thanks @winklerand

jreback commented on pandas/core/resample.py:
```diff
@@ -1243,6 +1214,51 @@ def _get_time_period_bins(self, ax):
+        bins = memb.searchsorted(rng, side='left')
+
+        if nat_count > 0:
+            # NaT handling as in pandas._lib.lib.generate_bins_dt64()
```
is this path tested sufficiently, e.g. 0, 1, 2 NaT?

winklerand replied:
added a test case for consecutive NaTs in the index (1cad7fa)

Should be sufficiently tested, cases covered:
- 0 NaT: basically all other resampling tests
- multiple single NaTs (at beginning, inside and end of index)
- consecutive NaTs (at beginning, inside and end of index)

Any ideas for more exhaustive test cases?

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -696,6 +696,32 @@ def create_index(self, *args, **kwargs):
         factory = self._index_factory()
         return factory(*args, **kwargs)
 
+    _index_fixture_start = datetime(2005, 1, 1)
```
as a followup PR. test_resample is getting sufficiently long that it makes sense to split this into some sub-dirs e.g.

``pandas/tests/resample/test_api, test_grouping, test_datetimeindex, test_period`` etc. IOW split out the classes to separate files (to start).

jreback replied:
we would normally co-locate fixtures into a ``common.py``

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -696,6 +696,32 @@ def create_index(self, *args, **kwargs):
+    _index_fixture_end = datetime(2005, 1, 10)
+    _index_fixture_freq = 'D'
+
+    @pytest.fixture(scope='class')
```
normally don't use scoping like this (and instead use default function scoping). The issue is that we *could* have mutating things which will then propogate (we shouldn't in any event).

winklerand replied:
makes sense, changed to default function scoping (6fe6c0c)

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -696,6 +696,32 @@ def create_index(self, *args, **kwargs):
 
+    _index_fixture_start = datetime(2005, 1, 1)
+    _index_fixture_end = datetime(2005, 1, 10)
+    _index_fixture_freq = 'D'
```
these can be fixtures themselves

winklerand replied:
done

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2167,58 +2194,47 @@ def test_resample_datetime_values(self):
+    @pytest.mark.parametrize('freq', ['2D'])
+    @pytest.mark.parametrize('kind', ['period', None, 'timestamp'])
+    def test_asfreq_downsample(self, series_and_frame, freq, kind):
+        # GH 12884, 15944
```
give a 1-liner about the issue that you are specifiying test when not obvious (e.g. here)

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2167,58 +2194,47 @@ def test_resample_datetime_values(self):
+    def test_asfreq_downsample(self, series_and_frame, freq, kind):
+        # GH 12884, 15944
+        obj = series_and_frame
+        start = obj.index[0].to_timestamp(how='start')
```
start/end only used for non-timestam, so put it there

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2167,58 +2194,47 @@ def test_resample_datetime_values(self):
+    def test_asfreq_upsample(self, series_and_frame, freq, kind):
+        # GH 12884, 15944
+        obj = series_and_frame
+        start = obj.index[0].to_timestamp(how='start')
```
same

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2167,58 +2194,47 @@ def test_resample_datetime_values(self):
+        start = obj.index[0].to_timestamp(how='start')
+        end = (obj.index[-1] + 1).to_timestamp(how='start')
+        if kind == 'timestamp':
+            expected = obj.to_timestamp().resample(freq).asfreq()
```
can this test be combined with prior test? (looks like you can just add to the freq), or is there a reason not?

winklerand replied:
yes, parametrization makes this obvious. In fact, both tests cover almost the same code path, so I consolidated them into one (f8f4157). 
Reason basically was not to change the existing structure - the same methods are defined in the ```Base``` class. I consolidated them as well in the ```Base``` class - PTAL.

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2772,6 +2788,71 @@ def test_evenly_divisible_with_no_extra_bins(self):
+        expected_index = self.create_index(df.index[0],
+                                           periods=len(df.index) / 2,
+                                           freq='2D')
+        # loffset coreces PeriodIndex to DateTimeIndex
```
blank line

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2772,6 +2788,71 @@ def test_evenly_divisible_with_no_extra_bins(self):
+            result_how = df.resample('2D', how=agg_arg, loffset='2H',
+                                     kind=kind)
+        if isinstance(agg_arg, list):
+            expected.columns = pd.MultiIndex .from_tuples([('value', 'mean')])
```
you have an odd space after pd.MultiIndex

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2772,6 +2788,71 @@ def test_evenly_divisible_with_no_extra_bins(self):
+        pi = PeriodIndex(start='2000', freq='D', periods=10)
+        s = Series(range(len(pi)), index=pi)
+        expected = s.to_timestamp().resample(freq).ohlc().to_period(freq)
+        # timestamp-based resampling doesn't include all sub-periods
```
blank line

winklerand replied:
done

jreback commented on pandas/tests/test_resample.py:
```diff
@@ -2772,6 +2788,71 @@ def test_evenly_divisible_with_no_extra_bins(self):
+        assert_frame_equal(result, expected)
+
+    def test_resample_with_only_nat(self):
+        pi = PeriodIndex([pd.NaT] * 3, freq='S')
```
issue number for this?

jreback reviewed (changes requested):
really just some minor stylistic changes.

can you fully close all of the mentioned issues at the top of the PR, and then create a *new* issue with any remnants. (ideally with check boxes). 
Also pls create a new issue for test_resample reorg as I suggested. thanks.

jreback commented:
> Whether or not we reveal kind on resamples is an open question - I think there already enough tools for switching between DateTimeIndex and PeriodIndex that we don't need more in resample.

For now this is fine. Yes I think we should remove this as PI becomes more flexible. But that's for another time.

jreback commented:
@winklerand biggest thing is going to need a small section in 0.21.0 to show these changes. In fact I find that writing these examples forces organization on describing the actual changes. Then (in this issue or followup), ideally like to example the timeseries.rst docs w.r.t. PI resampling.

winklerand committed:
STYLE: added blank lines, removed odd whitespace, fixed typo

winklerand committed:
TST: add test case for multiple consecutive NaTs in PeriodIndex

winklerand committed:
TST/DOC: added issue number to test case

winklerand committed:
TST: consolidate test_asfreq_downsample, test_asfreq_upsample -> test_asfreq

winklerand committed:
TST: set fixtures to default function scoping

winklerand committed:
TST: convert constant 'setup-like' values/objects to pytest fixtures

winklerand committed:
DOC: whatsnew v0.21.0 entry (in API changes section)

winklerand commented on pandas/tests/test_resample.py:
```diff
@@ -722,35 +722,19 @@ def series_and_frame(self, request, index):
-        expected = s.reindex(new_index)
-        assert_series_equal(result, expected)
+    @pytest.mark.parametrize('freq', ['2D', '1H'])
+    def test_asfreq(self, series_and_frame, freq):
```
@jreback 
I consolidated the ```test_asfreq_downsample()``` and ```test_asfreq_upsample()``` tests in the ```Base``` class as well (so we don't have to override them in ```TestPeriodIndex```). PTAL if everything's still correct and covered as before.

winklerand commented:
@jreback sorry for the delay on this one.

I think I made the requested changes, PTAL. Thanks for the great suggestions.

I put a small PeriodIndex resampling paragraph into the v0.21 whatsnew documentation, in the 'Backward incompatible API changes' section - is this the right place?

Going to organise the issues (closing the solved ones, open new one with the remaining issues) and create a test_resample reorg issue later tonight.

gfyoung commented:
@jreback : Were we waiting for anything on this one?

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -47,6 +47,87 @@ Backwards incompatible API changes
+
+.. code-block:: ipython
+
+   In [1]: pi = pd.period_range('2017-01', periods=12, freq='M')
```
the fixed things that are not changed (e.g. [1], [2]), do in a separate ipython block above (to avoid repeating in previous/new).

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -47,6 +47,87 @@ Backwards incompatible API changes
+``PeriodIndex`` resampling
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In previous versions of pandas, resampling a ``Series``/``DataFrame`` indexed by a ``PeriodIndex`` returned a ``DatetimeIndex`` in some cases (:issue:`12884`). Resampling to a multiplied frequency now returns a ``PeriodIndex`` (:issue:`15944`).
```
is this the only issue?

jreback commented on pandas/core/resample.py:
```diff
@@ -784,53 +784,32 @@ class PeriodIndexResampler(DatetimeIndexResampler):
+
+        if self.loffset is not None:
+            # Cannot apply loffset/timedelta to PeriodIndex -> convert to
+            # timestamps
```
should we show a warning for this?

gfyoung replied:
I think we should, given that we kind of did that before (it was `print` statement with a warning!?)

jreback reviewed (changes requested):
looks pretty good. couple of comments.

jreback commented:
side issue, maybe we should split up test_resample.py similar to groupby, e.g. make a sub-dir with common, test_datetime, test_period, test_timedelta. to make this a bit more understandable (let's make a separate issue for that).

jreback commented:
can you rebase. this looked really good. need to review once more.

jreback commented:
@winklerand sorry let this linger. can you rebase.

jreback committed:
fixups

jreback commented:
ok I rebased, will look after passing

jreback commented:
@MaximilianR can you have a look. in particular can you look at the original issues and see if we are fully / partially covering and make a note.

MaximilianR commented on pandas/core/resample.py:
```diff
@@ -1293,6 +1264,51 @@ def _get_time_period_bins(self, ax):
+
+        # if index contains no valid (non-NaT) values, return empty index
+        if not len(memb):
+            binner = labels = PeriodIndex(
```
You could use `_shallow_copy` here, but this is OK

jreback replied:
done

MaximilianR commented on pandas/tests/test_resample.py:
```diff
@@ -2812,18 +2817,96 @@ def test_evenly_divisible_with_no_extra_bins(self):
+    @pytest.mark.parametrize('agg_arg', ['mean', {'value': 'mean'}, ['mean']])
+    def test_loffset_returns_datetimeindex(self, frame, kind, agg_arg):
+        # make sure passing loffset returns DatetimeIndex in all cases
+        # basic method taken from Base.test_resample_loffset_arg_type()
```
This is OK-but-not-ideal behavior - I think pandas' approach is to write tests around that behavior but also to note that. 
Something like "TODO: not ideal to convert to DatetimeIndex, but...", and maybe an issue number

jreback replied:
this is ok for now, IIRC this is on the master PR list already?

MaximilianR commented:
This is superb, thank you @winklerand ! Much better functionality, and v nice & readable code & tests.

I still think the `kind=` kwarg is adds complication and reduces clarity, but that's for another day / the move to `IntervalIndex`.

Great work!

jreback committed:
moar whatsnew

jreback commented:
thanks @winklerand nice fixes!

winklerand commented:
thanks @jreback for pushing this one over the finish line, completely forgot about it (works for me syndrome ;-)), sorry!
